src/interface/trsm.hpp

```diff
@@ -0,0 +1,432 @@
+    invertBlocksEvent =
+        ex.execute(diagInverter, localSize, globalSize, localMemSize);
+  }
+
+  // Creates a copy of B to avoid overwriting the input in GEMM. While computing
+  // output X will hold the TRSM result and will be copied to B at the end
+  const index_t BSize = ldb * (N - 1) + M;
+  const index_t ldx = ldb;
+  auto X = make_sycl_iterator_buffer<element_t>(BSize);
+  auto bufferX = make_matrix_view<col_major>(ex, X, M, N, ldx);
+  trsmEvents = concatenate_vectors(
+      trsmEvents, ex.execute(make_op<Assign>(bufferX, bufferB), BSize));
+
+  if (isLeft) {
+    if ((isUpper && isTranspose) || (!isUpper && !isTranspose)) {
+      // Solves the system AX = alpha*B, as described in the documentation of
+      // the function when X is lower triangular
+
+      // True when (lower triangular) or (upper triangular and transposed)
+      for (index_t i = 0; i < M; i += blockSize) {
+        const index_t currentBlockSize = std::min(M - i, blockSize);
+        index_t gemmM = currentBlockSize;
+        index_t gemmN = N;
+        index_t gemmK = currentBlockSize;
+        char gemmTransA = isTranspose ? 't' : 'n';
+        char gemmTransB = 'n';
+        element_t gemmAlpha = (i == 0) ? alpha : element_t{1};
+        element_t gemmBeta = element_t{0};
+        index_t gemmLda = blockSize;
+        index_t gemmLdb = ldb;
+        index_t gemmLdc = ldx;
+        std::ptrdiff_t offsetA = i * blockSize;
+        std::ptrdiff_t offsetB = i;
+        std::ptrdiff_t offsetC = i;
+        auto gemmBufferA = invA + offsetA;
+        auto gemmBufferB = B + offsetB;
+        auto gemmBufferC = X + offsetC;
+        auto gemmEvent =
+            internal::_gemm(ex, gemmTransA, gemmTransB, gemmM, gemmN, gemmK,
+                            gemmAlpha, gemmBufferA, gemmLda, gemmBufferB,
+                            gemmLdb, gemmBeta, gemmBufferC, gemmLdc);
+        trsmEvents = concatenate_vectors(trsmEvents, gemmEvent);
+
+        if ((i + blockSize) >= M) {
+          break;
+        }
+
+        gemmM = M - i - blockSize;
+        gemmN = N;
+        gemmK = blockSize;
+        gemmTransA = isTranspose ? 't' : 'n';
+        gemmTransB = 'n';
+        gemmAlpha = element_t{-1};
+        gemmBeta = (i == 0) ? alpha : element_t{1};
+        gemmLda = lda;
+        gemmLdb = ldx;
+        gemmLdc = ldb;
+        offsetA = !isTranspose ? ((i + blockSize) + (i * lda))
+                               : (i + (blockSize + i) * lda);
+        offsetB = i;
+        offsetC = i + blockSize;
+        gemmBufferA = A + offsetA;
+        gemmBufferB = X + offsetB;
+        gemmBufferC = B + offsetC;
+        gemmEvent =
+            internal::_gemm(ex, gemmTransA, gemmTransB, gemmM, gemmN, gemmK,
+                            gemmAlpha, gemmBufferA, gemmLda, gemmBufferB,
+                            gemmLdb, gemmBeta, gemmBufferC, gemmLdc);
+        trsmEvents = concatenate_vectors(trsmEvents, gemmEvent);
```

Comment:
same as above. These operations are repeated multiple times. Can you wrap them in a function and call the function instead?

Reply:
They are repeated multiple times but with different parameters